primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,397 @@
+	pub fn unlock(&mut self) {
+		self.value_ref.clear();
+	}
+}
+
+/// Extension trait for locks based on [`Timestamp`](::sp_core::offchain::Timestamp).
+///
+/// Allows explicity setting the timeout on construction
+/// instead of using the implicit default timeout of
+/// [`STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS`](Self::STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS).
+pub trait TimeLock<'a>: Sized {
+	/// Provide an explicit deadline timestamp at which the locked state of the lock
+	/// becomes stale and may be dismissed by `fn try_lock(..)`, `fn spin_lock(..)` and others.
+	fn with_deadline<'k>(key: &'k [u8], lock_deadline: Timestamp) -> Self
+	where
+		'k: 'a;
+}
+
+impl<'a> TimeLock<'a> for StorageLock<'a, Timestamp> {
+	fn with_deadline<'k>(key: &'k [u8], lock_deadline: Timestamp) -> Self
+	where
+		'k: 'a,
+	{
+		Self {
+			value_ref: StorageValueRef::<'a>::persistent(key),
+			deadline: lock_deadline,
+		}
+	}
+}
+
+/// Bound for block numbers which commonly will be implemented by the `frame_system::Trait::BlockNumber`.
+///
+/// This trait has no intrinsic meaning and exists only to decouple `frame_system`
+/// from `runtime` crate and avoid a circular dependency.
+pub trait BlockNumberTrait: Codec + Copy + Clone + Ord + Eq {
+	/// Returns the current block number.
+	///
+	/// Commonly this will be implemented as
+	/// ```rust
+	/// fn current() -> Self {
+	///     frame_system::Module<Trait>::block_number()
+	/// }
+	/// ```
+	/// but note that the definition of current is
+	/// application specific.
+	fn current() -> Self;
+}
+
+/// Extension for lor storage locks which are based on blocks in addition to a timestamp.
+trait BlockAndTimeLock<'a, B>: Sized
+where
+	B: BlockNumberTrait,
+{
+	fn with_block_and_time_deadline<'k>(
+		key: &'k [u8],
+		block_deadline: B,
+		time_deadline: Timestamp,
+	) -> Self
+	where
+		'k: 'a;
+}
+
+impl<'a, B> BlockAndTimeLock<'a, B> for StorageLock<'a, BlockAndTime<B>>
+where
+	B: BlockNumberTrait,
+{
+	fn with_block_and_time_deadline<'k>(
+		key: &'k [u8],
+		block_deadline: B,
+		time_deadline: Timestamp,
+	) -> Self
+	where
+		'k: 'a,
+	{
+		Self {
+			value_ref: StorageValueRef::<'a>::persistent(key),
+			deadline: BlockAndTime {
+				block_number: block_deadline,
+				timestamp: time_deadline,
+			},
+		}
+	}
+}
+
+/// RAII style guard for a lock.
+pub struct StorageLockGuard<'a, 'b, L>
+where
+	L: Lockable,
+{
+	lock: Option<&'b mut StorageLock<'a, L>>,
+}
+
+impl<'a, 'b, L> StorageLockGuard<'a, 'b, L>
+where
+	L: Lockable,
+{
+	/// Consume the guard but DO NOT unlock the underlying lock.
+	pub fn forget(mut self) {
+		let _ = self.lock.take();
+	}
+}
+
+impl<'a, 'b, L> Drop for StorageLockGuard<'a, 'b, L>
+where
+	L: Lockable,
+{
+	fn drop(&mut self) {
+		if let Some(lock) = self.lock.take() {
+			lock.unlock();
+		}
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use sp_core::offchain::{testing, OffchainExt, OffchainStorage};
+	use sp_io::TestExternalities;
+
+	#[test]
```

Comment:
I think this amount of great work deserves more than one test 😄

Reply:
Requires a new `Sync` impl of `TestOffchainExt`, otherwise all tests are bogus :)

Reply:
Added a few more tests, but concurrent access tests are pending.